File: edge-worker/src/types/link-record.ts
/**
 * AutoFix — LinkRecord (L2 shared state)
 * Sourced from autofix-polyglot types/typescript/link-record.ts
 * Source of truth: schemas/link-record.schema.json
 */

export type LinkStatus = "PENDING" | "HEALED" | "DEAD" | "HEALTHY";

export interface LinkRecord {
  status: LinkStatus;
  original_url: string;
  resolved_url?: string;
  discovered_at?: string; // ISO-8601
  healed_at?: string; // ISO-8601
  reason?: string;
}

/** Canonical KV key for an absolute URL (must match Go types.KeyFor). */
export function urlKey(url: string): string {
  if (typeof TextEncoder !== "undefined") {
    const bytes = new TextEncoder().encode(url);
    let binary = "";
    for (let i = 0; i < bytes.length; i++) {
      binary += String.fromCharCode(bytes[i]!);
    }
    return btoa(binary);
  }
  return btoa(unescape(encodeURIComponent(url)));
}

File: healer/types/linkrecord.go
// Package types holds shared AutoFix contracts.
// Sourced from autofix-polyglot types/go/linkrecord.go
// Keep structs aligned with JSON Schemas and TypeScript types.
package types

import (
	"encoding/base64"
)

// LinkStatus values mirror schemas/link-record.schema.json enum.
const (
	StatusPending = "PENDING"
	StatusHealed  = "HEALED"
	StatusDead    = "DEAD"
	StatusHealthy = "HEALTHY"
)

// LinkRecord is the shared KV value (L2).
type LinkRecord struct {
	Status       string `json:"status"`
	OriginalURL  string `json:"original_url"`
	ResolvedURL  string `json:"resolved_url,omitempty"`
	DiscoveredAt string `json:"discovered_at,omitempty"`
	HealedAt     string `json:"healed_at,omitempty"`
	Reason       string `json:"reason,omitempty"`
}

// DiscoveryMessage is the queue payload (Edge → Queue).
type DiscoveryMessage struct {
	URLs         []string `json:"urls"`
	DiscoveredAt string   `json:"discovered_at"`
}

// DiscoverRequest is the body of POST /v1/discover.
type DiscoverRequest struct {
	URLs []string `json:"urls,omitempty"`
	URL  string   `json:"url,omitempty"`
}

// DiscoverResponse is the success body of POST /v1/discover.
type DiscoverResponse struct {
	Enqueued int `json:"enqueued"`
}

// KeyFor returns the canonical KV key for an absolute URL.
// Must match TypeScript urlKey() (UTF-8 bytes → Std base64).
func KeyFor(url string) string {
	return base64.StdEncoding.EncodeToString([]byte(url))
}
